### DISCUSSION TITLE
input_timetable

### DISCUSSION THREAD

**Discussion by amitfin** [home-assistant/architecture#751] [Entity Models]
# The Need
Having an automation rule's input element with a list of time events.
(input_datetime can specify only a single event).
# The Proposal
A new input element - `input_timetable`:
- The element has a "timetable" attribute which contains a list of pairs <time, state>.
- The state of the element is determined by the above list.
- The automation rule uses the state of the new element in its logic.
# Additional Details
A short demo clip can be found [here](https://www.youtube.com/watch?v=TluMAZpORwk).

The PRs:
- [Core](https://github.com/home-assistant/core/pull/41385)
- [Doc](https://github.com/home-assistant/home-assistant.io/pull/15023)
- [Frontend1](https://github.com/home-assistant/frontend/pull/7629) (more-info dialog)
- [Frontend2](https://github.com/home-assistant/frontend/pull/10687) (new helper)

A post with a description of an end-to-end project using the new element can be found [here](https://community.home-assistant.io/t/automating-an-entire-day/398731).

Thoughts?

**Comment by balloob** [comment]
Time triggers [allow multiple values](https://www.home-assistant.io/docs/automation/trigger#multiple-times), including multiple entities

Wouldn't it make sense to add support for calendar entities to time trigger instead? It pulls in a calendar and so it can have multiple events.

**Reply by allenporter** [reply]
Related also is a discussion about adding triggers for Calendar:
https://github.com/home-assistant/architecture/discussions/700#discussioncomment-1902322

I'm not quite sure if these are the same thing or different, since I haven't looked closely at time.

A timetable does seem like a calendar to me. Is this proposal also filling a gap in available calendar integrations? (e.g. local calendar has been floated before i believe)

**Reply by balloob** [reply]
It sounds powerful, you can have either a local calendar but also be able to use any calendar source for your timetable.

And you can use recurring events so you don't need to add everything every week and can have exceptions for holidays etc. You can use the UI on your phone etc.

**Reply by amitfin** [reply]
> Time triggers [allow multiple values](https://www.home-assistant.io/docs/automation/trigger#multiple-times), including multiple entities

Manipulating automation rule's time triggers requires editing the rule itself. The purpose of `input_timetable` is to keep the time manipulation outside of the rule config (like the other "input" elements). 

> Wouldn't it make sense to add support for calendar entities to time trigger instead? It pulls in a calendar and so it can have multiple events.

`input_timetable` has on and off state. Keeping it in the time trigger section will mean maintaining 2 set of rules (e.g. one for turning on the lights and another one for off).

**Reply by balloob** [reply]
Calendar events have start and end and so an event is "active" or not (`on`/`off` state).

I really do not see a difference between timetable and a calendar and I suggest we focus our efforts on improving using the calendar integration for automation triggers.

Using calendar integration means you can use any calendar integration (including a local one if created) and either make a time table (weekly recurring events) or have a more sporadic schedule like package deliveries.

**Reply by amitfin** [reply]
Does calendar mean [Google Calendars integration](https://www.home-assistant.io/integrations/google/)?
We can change the name of this integration to "local calendar" (or input_calendar) to align the two, unless I'm missing something more fundamental here.

**Reply by balloob** [reply]
I am talking about the core integration [`calendar`](https://www.home-assistant.io/integrations/calendar) that is how integrations like CalDav and Google integrate.

It would require we add a local calendar variant that adheres to RFC5545. A user can create recurring weekly events for a weekly schedule, or any other schedule. They can do that locally, or in any of the other calendar sources.

Then having calendar triggers would unlock all possibilities for all integrations instead of just one.

**Reply by amitfin** [reply]
RFC5545 brings more flexibility, but it's a different size of a project.
`input_timetable` is a helper element which lives completely inside the UI of HA. It can support many use cases which requires a list of time events.
Maybe we can look on both approaches as complementary: in many cases a simple helper will do it. Otherwise, a Calendar integration should be used.

**Reply by balloob** [reply]
I feel like we both want the same: an easy UI in the automation editor to set a schedule for when the automation needs to trigger (start/stop events). Having to duplicate this logic and UI for both calendar and input_timetable will lead to a confusing user experience with 2 ways to do the same thing.

The problem is that a simple helper never stay simple. People want to add more features all the time and we will end up rebuilding the calendar integration. That's why I propose to just make the step right away to support calendar integration. 

And having a calendar trigger using a local calendar could still provide a simple editing UI in the automation editor.

**Reply by Jc2k** [reply]
So would a "local calendar" have its UI inside HA core? Or would all calendars gain a UI for managing a schedule directly in HA?

**Reply by amitfin** [reply]
> The problem is that a simple helper never stay simple. People want to add more features all the time and we will end up rebuilding the calendar integration. That's why I propose to just make the step right away to support calendar integration.

This is the root decision here: should we add a simplified way to manage list of time events with the following 2 considerations:
1) Having a full blown local calendar is a superior solution (which might happen or not)
2) Adding more features to the new helper might make it less "simplified"

To make it more concrete, here is a short clip which demonstrate `input_datetime` vs. `intput_timetable`. One is the plural version of the other:

https://user-images.githubusercontent.com/19599059/157320379-65c79926-38b0-4442-a945-9067e83a0a81.mov

**Reply by balloob** [reply]
@Jc2k we already have a full calendar UI in HA to show calendar events. We don't use it yet for creation. 

@amitfin that UI is not clear what is "start" and "stop". It's only start. I thought you mentioned you cared about both start and stop.

**Reply by allenporter** [reply]
Google calendar has a create event service if we want to standardize on it and try this UI without waiting for a local integration implementation.

**Reply by amitfin** [reply]
> @amitfin that UI is not clear what is "start" and "stop". It's only start. I thought you mentioned you cared about both start and stop.

The toggle on the right controls the `on`/`off` value. (Not a UI expert, so open for feedback 😊)
<img width="410" alt="image" src="https://user-images.githubusercontent.com/19599059/157402016-2af489b9-d817-451a-b2f9-a66c7da77495.png">
<img width="418" alt="image" src="https://user-images.githubusercontent.com/19599059/157402270-16f366c5-35cc-4c5b-bb4d-96c480079880.png">

**Reply by amitfin** [reply]
Here is an alternative design to make the `state` property clearer:
![image](https://user-images.githubusercontent.com/19599059/157468872-d9b0d6be-4b79-4330-ba7d-2134527691e5.png)

**Reply by allenporter** [reply]
I was reviewing the old PRs in the backlog and came across https://github.com/home-assistant/core/pull/41385 which brought me back to this discussion. After doing a bunch of improvements on google calendar and calendar automations, i've been thinking a lot about this thread, and more about a local calendar would look like.

I'm seeing this break down into these topics for further discussion/exploration and curious to hear thoughts.

## Entity service for event creation

Google Calendar currently has a [create event service](https://github.com/home-assistant/core/blob/d3f4108a91bfb8c7607e145ab48cfd489143d972/homeassistant/components/google/services.yaml#L55) that is fairly straight forward, and can be pushed up to the calendar integration, and delegated to an optional new entity method on `CalendarEntity`.

Other eligible integrations for this can be  `caldav` and the python caldav client library looks like it supports calendar event creation.

## Recurring events

The [google calendar api](https://developers.google.com/calendar/api/guides/recurringevents) and [caldav client example](https://github.com/python-caldav/caldav/blob/6b791de7ccae48901ada8318158ae824f8f36982/examples/basic_usage_examples.py#L66)) both support recurring events via `RRULE` so that is nice.

## Editing Events

Editing calendar events seems pretty important, and also significantly increases the complexity. We'd need to introduce event ids, and  likely effectively support partial edits. A more detailed editing proposal is probably a blocker before proceeding with this

## User Interface

Presumably a lot of the great learnings from the input timetable can be applied to a frontend calendar UI creator and editor? I'm not sure how easy it would be to leverage data options/config flow based UIs for this.

## Local Calendar support

 I've been looking at  ics-py which seems like the most active python rfc5545 implementation.  It is missing support for recurring events which is the key thing you need for an input timetable ([issue open](https://github.com/ics-py/ics-py/issues/14) for 5+ years for example).

I was chatting with @balloob about this topic in person and he seemed to have a lot of personal experience dealing with this complexity of calendar rfc5545. He suggested another route is to lean into a local caldav server as an alternative which i think is a good as another path to explore to give us more near term options. (e.g. there may already be a popular lightweight caldav add-on server that can be recommended and streamlined for example)

**Reply by frenck** [reply]
> After doing a bunch of improvements on google calendar and calendar automations, i've been thinking a lot about this thread, and more about a local calendar would look like.

We had quite an extensive discussion (chat) about this all with a couple of people in the past weeks. And made some conclusion on moving the original PR and vision forward. I've collected them and working on the proposal based on those conclusions. 

It would allow both a timetable + local calendar. I will provide this in the next couple of days. I've been putting it aside as we've had some CI issues that needed to be fixed, and right now working to get a beta shipped.

**Comment by thecode** [comment]
I think that the suggested time table helper is very useful, maybe it also needs an option to select week days, but I don't think it is similar to a calendar.
Currently I have few automations which run at specific times, but sometime I want to disable a single occurrence and enable it back later.
A good example is watering system, this should run at fixed schedules, regardless of holidays or special events. 
However when it is hotter or colder I sometime add or remove time to the watering system. Currently I do the scheduling in node-red since I did not find a simple solution which I can easily change the schedule. Calendar based triggers may work but seems to complicated to use for a fixed schedules automations.

**Reply by allenporter** [reply]
Interesting. I actually run my lights on a fixed schedule with a calendar, but then have additional conditions/triggers that are season like sun/sunset in the mix as well.  My sprinkler uses a calendar, but it's denormalized from some high level guidance like "every 3 days", and can take inputs like "delay by 2 days" because it rained to shit the calendar based time table.

I'm not sure if we are talking about differences between how we like to configure our integrations or input data (style or workflow preferences), or if there are underlying fundamental differences in how the apis look for providing data.

**Reply by thecode** [reply]
> Interesting. I actually run my lights on a fixed schedule with a calendar, but then have additional conditions/triggers that are season like sun/sunset in the mix as well. My sprinkler uses a calendar, but it's denormalized from some high level guidance like "every 3 days", and can take inputs like "delay by 2 days" because it rained to shit the calendar based time table.
> 
> I'm not sure if we are talking about differences between how we like to configure our integrations or input data (style or workflow preferences), or if there are underlying fundamental differences in how the apis look for providing data.

Can you show how it is configured from the UI? Just to understand if I am missing something because the main benefit for me is an option to adjust schedules with simple UI interface.

**Reply by amitfin** [reply]
> ... the main benefit for me is an option to adjust schedules with simple UI interface

This is the main purpose of this integration. Thanks!

**Reply by allenporter** [reply]
> Can you show how it is configured from the UI? Just to understand if I am missing something because the main benefit for me is an option to adjust schedules with simple UI interface.

Apologies, I gave the wrong impression: what i mean to say for the sprinkler is that its another product outside of home assistant.  You [drag sliders](https://wifi.rainbird.com/wp-content/uploads/2020/12/Screenshot-2020-12-30-at-10.48.26-AM-473x1024.jpg) and it shows you the outcomes on a calendar as a schedule (not shown, but its how you might expect)  I'm really wondering about is if we can separate the data layer and APIs from the user perspective, and if that helps show one way or the other if the requirements are similar or different.

**Comment by jeeftor** [comment]
So I mocked up this blueprint last year: https://community.home-assistant.io/t/wip-day-time-selector/333870?u=jeeftor I never fully finished it - but is this similar to what you are wanting?


![](https://community-assets.home-assistant.io/original/3X/1/5/159ac899e52027e6ff5fea76026cb1c6de14b767.png)

**Reply by amitfin** [reply]
Thanks!
What I was trying to build is a variable list of time events (start and end are 2 such events). The goal is to be able to manipulate easily such a list (adding / removing / editing entries), without the need to change the automation rule itself.

**Reply by jeeftor** [reply]
Basically an improved version of my blueprint that is directly integrated with the UI? Sounds great!


And it sounds like you want the ability to enable/disable the various events?


At 8AM - Monday, Wednesday, Thursday Do Something  (Turn on sprinkler)
At 8:05AM - Monday, Wednesday, Thursday Do something else  (Turn off Sprinkler)

And then you get 6 events out of this?

- `T1` (Monday_ON)
- `T2` (Monday_OFF)
- `T3` (Wednesday_On)
- `T4` (Wednesday_Off)
- `T5` (Thursday_On)
- `T6` (Thursday_Off)

And you want the ability to "disable" say `T5` based on some condition?

**Reply by amitfin** [reply]
In our use case, we want to easily add additional time events (e.g. turn on and off the Sprinkler not only in the morning but also in the evening). 
It's possible to create "standby" time events and enable / disable them as needed. However, a more natural way is to support a list of events (with add / remove / edit functionality). `input_timetable` is the plural version of `input_datetime`.
Does it make sense?

**Comment by frenck** [comment]
We are a year later, we now have `schedule` and `calendar` integrations to handle the cases.

Closing this proposal for that reason.

../Frenck
